AdminClient/src/app/customers/customers.component.spec.ts

```diff
@@ -22,4 +22,45 @@ describe('CustomersComponent', () => {
   it('should create', () => {
     expect(component).toBeTruthy();
   });
+
+  it('should contain column text in the table',() => {
+    const compiled = fixture.debugElement.nativeElement;
+    expect(compiled.querySelector('#table').textContent).toContain('ID' && 'Name' && 'Category'
+    && 'Telephone' && 'Email' && 'Website' && 'Cont' && 'Address' && 'Town/City' && 'County' && 'Country' &&
+      'Postcode');
+  });
+
+
+  it('should add a customer to a customer collection', () => {
+    component.customers = [];
+    let customer = {id: 1, name: 'a', category: 'a', telephone: 'a', email: 'a',
+        website: 'a', contact: 'a', address: 'a', townOrCity: 'a',
+        county: 'a', country: 'a', postcode: 'a'};
+    component.customers.push(customer);
+    customer = {id: 2, name: 'b', category: 'b', telephone: 'b', email: 'b',
+        website: 'b', contact: 'b', address: 'b', townOrCity: 'b',
+        county: 'b', country: 'b', postcode: 'b'};
+    component.customers.push(customer);
+    expect(component.customers.length).toBe(2);
+  });
+
+  it('should remove a customer from a customer collection', () => {
+    component.customers = [{id: 1, name: 'a', category: 'a', telephone: 'a', email: 'a',
+      website: 'a', contact: 'a', address: 'a', townOrCity: 'a',
+      county: 'a', country: 'a', postcode: 'a'}];
+    component.customers.pop();
+    expect(component.customers.length).toBe(0);
+  });
+
+  it('should load dialog data (dialog type)', () => {
+    const inputData = 'load';
+    component.data = inputData;
+    expect(component.data).toBe('load');
+  });
+
+  it('should call the update record dialog once', () => {
+    spyOn(component, 'openUpdateRecordDialog')
+    component.openUpdateRecordDialog()
+    expect(component.openUpdateRecordDialog).toHaveBeenCalledTimes(1)
+  });
```

Comment:
```suggestion

```
This was a good attempt in introducing unit tests however some things to take into consideration.

For example
```
  component.customers = [{id: 1, name: 'a', category: 'a', telephone: 'a', email: 'a',
      website: 'a', contact: 'a', address: 'a', townOrCity: 'a',
      county: 'a', country: 'a', postcode: 'a'}];
    component.customers.pop();
```

What you are testing here is that JavaScript is working that is expected. You simply add an element to the array then use the **array** function to remove the element.

What we want to do when testing is change the state whether that is in our app or a more general state i.e NgRx and then check if for example the HTML has changed to reflect that change. Or something else we could be testing is that a custom function we made returns the expected result or triggers the subsequent actions that must follow.

Tl;dr what you are doing here is testing that JavaScript works.

Reply:
Yes sorry, I was waiting for your table changes so I can check the resulting HTML. Will be added